### PR TITLE
ntokens_seen now accounts for padding tokens

### PR DESCRIPTION
**Pull request by francesco-bertolotti**
`self.ntokens_seen`, which is used downstream to report the total number of tokens seen during training, is currently computed using `labels.numel()`. This includes padding tokens.

At the same time, `local_valid_tokens`, which is used in the loss computation, is computed as `(labels != IGNORE_INDEX).sum()`, excluding padding tokens.

When the data contains no padding, these two counts are identical. However, they diverge when padding is present.

Since `ntokens_seen` is intended to represent the true number of tokens processed during training, it makes sense to compute it consistently with `local_valid_tokens`. This PR updates `self.ntokens_seen` to use the same padding-aware token count.

### CODE REVIEW

File: torchtitan/trainer.py
Comment (tianyu-l) [inline torchtitan/trainer.py:886]: > self.ntokens_seen, which is used downstream to report the total number of tokens seen during training, is currently computed using labels.numel(). This includes padding tokens.

I agree with this, but not
> When the data contains no padding, these two counts are identical.

In particular, for SFT the prompt tokens will be masked out by this label as well. See
https://github.com/pytorch/torchtitan/blob/main/torchtitan/hf_datasets/text_datasets.py#L498-L499

Comment (francesco-bertolotti) [reply torchtitan/trainer.py:886]: You're right, I wasn't accounting for SFT data. `IGNORE_INDEX` is written for two different reasons (padding and prompt masking) so `(labels != IGNORE_INDEX).sum()` is really "loss tokens", not "non-padding tokens".

So there are three candidate semantics for `ntokens_seen`:

1. loss tokens only -> `(labels != IGNORE_INDEX).sum()`
2. real tokens seen in the forward pass (loss tokens + masked prompt tokens, excluding padding) -> not super sure on whats the best way to compute this.
3. all positions — `labels.numel()` (current behavior)
